File: src/main/java/org/roborace/lapscounter/domain/MessageResult.java
package org.roborace.lapscounter.domain;

import lombok.*;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

@Getter
@Setter
@ToString
@EqualsAndHashCode
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class MessageResult {

    private List<Message> messages;
    private ResponseType responseType;

    private MessageResult(Message message, ResponseType responseType) {
        this.messages = Collections.singletonList(message);
        this.responseType = responseType;
    }

    public static MessageResult single(Message message) {
        return new MessageResult(message, ResponseType.SINGLE);
    }

    public static MessageResult broadcast(Message message) {
        return new MessageResult(message, ResponseType.BROADCAST);
    }

    public static MessageResult broadcast() {
        return new MessageResult(new ArrayList<>(), ResponseType.BROADCAST);
    }

    public void add(Message message) {
        messages.add(message);
    }

    public void addAll(List<Message> messages) {
        this.messages.addAll(messages);
    }
}

File: src/main/kotlin/org/roborace/lapscounter/domain/api/MessageResult.kt
package org.roborace.lapscounter.domain.api

import org.roborace.lapscounter.domain.api.ResponseType.BROADCAST
import org.roborace.lapscounter.domain.api.ResponseType.SINGLE


data class MessageResult(
    var responseType: ResponseType,
    var messages: MutableList<Message> = mutableListOf(),
) {
    fun add(message: Message) = messages.add(message)
    fun addAll(messages: List<Message>) = this.messages.addAll(messages)


    companion object {
        @JvmStatic
        fun single(message: Message) = MessageResult(SINGLE, mutableListOf(message))

        @JvmStatic
        fun single(messages: List<Message>) = MessageResult(SINGLE, messages.toMutableList())

        @JvmStatic
        fun broadcast(message: Message) = MessageResult(BROADCAST, mutableListOf(message))

        @JvmStatic
        fun broadcast(messages: List<Message>) = MessageResult(BROADCAST, messages.toMutableList())
    }
}
